File: redwood-yoga-java/src/main/java/io/github/orioncraftmc/meditate/internal/YGStyle.java
package io.github.orioncraftmc.meditate.internal;

import io.github.orioncraftmc.meditate.internal.detail.CompactValue;
import static io.github.orioncraftmc.meditate.internal.detail.GlobalMembers.bitWidthFn;
import static io.github.orioncraftmc.meditate.internal.detail.GlobalMembers.getEnumData;
import static io.github.orioncraftmc.meditate.internal.detail.GlobalMembers.setEnumData;
import io.github.orioncraftmc.meditate.internal.detail.Values;
import io.github.orioncraftmc.meditate.internal.enums.*;
import java.util.HashMap;
import java.util.Map;
import org.jetbrains.annotations.NotNull;

public class YGStyle //Type originates from: YGStyle.h
{

    private static final Integer directionOffset = 0;
    private static final Integer flexdirectionOffset = directionOffset + bitWidthFn(YGDirection.class);


    private static final Integer justifyContentOffset = flexdirectionOffset + bitWidthFn(YGFlexDirection.class);
    private static final Integer alignContentOffset = justifyContentOffset + bitWidthFn(YGJustify.class);

    //  ~YGStyle() = default;
    private static final Integer alignItemsOffset = alignContentOffset + bitWidthFn(YGAlign.class);
    private static final Integer alignSelfOffset = alignItemsOffset + bitWidthFn(YGAlign.class);
    private static final Integer positionTypeOffset = alignSelfOffset + bitWidthFn(YGAlign.class);
    private static final Integer flexWrapOffset = positionTypeOffset + bitWidthFn(YGPositionType.class);
    private static final Integer overflowOffset = flexWrapOffset + bitWidthFn(YGWrap.class);
    private static final Integer displayOffset = overflowOffset + bitWidthFn(YGOverflow.class);
    private final Values<YGEdge> margin_ = new Values<>();
    private final Values<YGEdge> position_ = new Values<>();
    private final Values<YGEdge> padding_ = new Values<>();
    private final Values<YGEdge> border_ = new Values<>();
    private final Values<YGDimension> dimensions_ = new Values<>(CompactValue.ofAuto().convertToYgValue());
    private final Values<YGDimension> minDimensions_ = new Values<>();
    private final Values<YGDimension> maxDimensions_ = new Values<>();
    private final Map<Object, Object> flags = new HashMap<>(Map.of());
    private YGFloatOptional aspectRatio_ = new YGFloatOptional();
    private YGFloatOptional flex_ = new YGFloatOptional();
    private YGFloatOptional flexGrow_ = new YGFloatOptional();
    private YGFloatOptional flexShrink_ = new YGFloatOptional();
    private CompactValue flexBasis_ = CompactValue.ofAuto();

    public YGStyle() {
        setEnumData(YGAlign.class, flags, alignContentOffset, YGAlign.YGAlignFlexStart);
        setEnumData(YGAlign.class, flags, alignItemsOffset, YGAlign.YGAlignStretch);
    }

    public void setAspectRatio(YGFloatOptional aspectRatio_) {
        this.aspectRatio_ = aspectRatio_;
    }

    public void setFlex(YGFloatOptional flex_) {
        this.flex_ = flex_;
    }

    public void setFlexGrow(YGFloatOptional flexGrow_) {
        this.flexGrow_ = flexGrow_;
    }

    public void setFlexShrink(YGFloatOptional flexShrink_) {
        this.flexShrink_ = flexShrink_;
    }

    public void setFlexBasis(CompactValue flexBasis_) {
        this.flexBasis_ = flexBasis_;
    }

    public final YGDirection direction() {
        return getEnumData(YGDirection.class, flags, directionOffset);
    }

    public final @NotNull BitfieldRef<YGDirection> directionBitfieldRef() {
        return new BitfieldRef<>(this, directionOffset);
    }

    public final YGFlexDirection flexDirection() {
        return getEnumData(YGFlexDirection.class, flags, flexdirectionOffset);
    }

    public final @NotNull BitfieldRef<YGFlexDirection> flexDirectionBitfieldRef() {
        return new BitfieldRef<>(this, flexdirectionOffset);
    }

    public final YGJustify justifyContent() {
        return getEnumData(YGJustify.class, flags, justifyContentOffset);
    }

    public final @NotNull BitfieldRef<YGJustify> justifyContentBitfieldRef() {
        return new BitfieldRef<>(this, justifyContentOffset);
    }

    public final YGAlign alignContent() {
        return getEnumData(YGAlign.class, flags, alignContentOffset);
    }

    public final @NotNull BitfieldRef<YGAlign> alignContentBitfieldRef() {
        return new BitfieldRef<>(this, alignContentOffset);
    }

    public final YGAlign alignItems() {
        return getEnumData(YGAlign.class, flags, alignItemsOffset);
    }

    public final @NotNull BitfieldRef<YGAlign> alignItemsBitfieldRef() {
        return new BitfieldRef<>(this, alignItemsOffset);
    }

    public final YGAlign alignSelf() {
        return getEnumData(YGAlign.class, flags, alignSelfOffset);
    }

    public final @NotNull BitfieldRef<YGAlign> alignSelfBitfieldRef() {
        return new BitfieldRef<>(this, alignSelfOffset);
    }

    public final YGPositionType positionType() {
        return getEnumData(YGPositionType.class, flags, positionTypeOffset);
    }

    public final @NotNull BitfieldRef<YGPositionType> positionTypeBitfieldRef() {
        return new BitfieldRef<>(this, positionTypeOffset);
    }

    public final YGWrap flexWrap() {
        return getEnumData(YGWrap.class, flags, flexWrapOffset);
    }

    public final @NotNull BitfieldRef<YGWrap> flexWrapBitfieldRef() {
        return new BitfieldRef<>(this, flexWrapOffset);
    }

    public final YGOverflow overflow() {
        return getEnumData(YGOverflow.class, flags, overflowOffset);
    }

    public final @NotNull BitfieldRef<YGOverflow> overflowBitfieldRef() {
        return new BitfieldRef<>(this, overflowOffset);
    }


    public final YGDisplay display() {
        return getEnumData(YGDisplay.class, flags, displayOffset);
    }

    public final @NotNull BitfieldRef<YGDisplay> displayBitfieldRef() {
        return new BitfieldRef<>(this, displayOffset);
    }


    public final YGFloatOptional flex() {
        return flex_;
    }

    public final YGFloatOptional flexGrow() {
        return flexGrow_;
    }

    public final YGFloatOptional flexShrink() {
        return flexShrink_;
    }

    public final CompactValue flexBasis() {
        return flexBasis_;
    }


    public final @NotNull Values<YGEdge> margin() {
        return margin_;
    }


    public final @NotNull Values<YGEdge> position() {
        return position_;
    }


    public final @NotNull Values<YGEdge> padding() {
        return padding_;
    }


    public final @NotNull Values<YGEdge> border() {
        return border_;
    }


    public final @NotNull Values<YGDimension> dimensions() {
        return dimensions_;
    }


    public final @NotNull Values<YGDimension> minDimensions() {
        return minDimensions_;
    }


    public final @NotNull Values<YGDimension> maxDimensions() {
        return maxDimensions_;
    }


    public final YGFloatOptional aspectRatio() {
        return aspectRatio_;
    }

    public static class BitfieldRef<T extends Enum<T>> //Type originates from: YGStyle.h
    {
        public final YGStyle style;
        public final Integer offset;

        public BitfieldRef(YGStyle style, Integer offset) {
            this.style = style;
            this.offset = offset;
        }

        public T getValue(@NotNull Class<T> enumClazz) {
            return getEnumData(enumClazz, style.flags, offset);
        }

        public final @NotNull BitfieldRef<T> setValue(@NotNull T x) {
            setEnumData((Class<T>) x.getClass(), style.flags, offset, x);
            return this;
        }
    }

}

File: redwood-yoga/src/commonMain/kotlin/app/cash/redwood/yoga/internal/YGStyle.kt
/*
 * Copyright (C) 2023 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:Suppress("unused")

package app.cash.redwood.yoga.internal

import app.cash.redwood.yoga.internal.detail.CompactValue
import app.cash.redwood.yoga.internal.detail.GlobalMembers
import app.cash.redwood.yoga.internal.detail.Values
import app.cash.redwood.yoga.internal.enums.YGAlign
import app.cash.redwood.yoga.internal.enums.YGDimension
import app.cash.redwood.yoga.internal.enums.YGDirection
import app.cash.redwood.yoga.internal.enums.YGDisplay
import app.cash.redwood.yoga.internal.enums.YGEdge
import app.cash.redwood.yoga.internal.enums.YGFlexDirection
import app.cash.redwood.yoga.internal.enums.YGJustify
import app.cash.redwood.yoga.internal.enums.YGOverflow
import app.cash.redwood.yoga.internal.enums.YGPositionType
import app.cash.redwood.yoga.internal.enums.YGWrap
import kotlin.reflect.KClass

internal class YGStyle {
  val margin = Values<YGEdge>()
  val position = Values<YGEdge>()
  val padding = Values<YGEdge>()
  val border = Values<YGEdge>()
  val dimensions = Values<YGDimension>(CompactValue.ofAuto().convertToYgValue())
  val minDimensions = Values<YGDimension>()
  val maxDimensions = Values<YGDimension>()
  var aspectRatio = YGFloatOptional()
  var flex = YGFloatOptional()
  var flexGrow = YGFloatOptional()
  var flexShrink = YGFloatOptional()
  var flexBasis = CompactValue.ofAuto()
  private val flags = mutableMapOf<Any?, Any>()

  init {
    GlobalMembers.setEnumData(
      YGAlign::class, flags, alignContentOffset, YGAlign.YGAlignFlexStart,
    )
    GlobalMembers.setEnumData(
      YGAlign::class, flags, alignItemsOffset, YGAlign.YGAlignStretch,
    )
  }

  fun direction(): YGDirection {
    return GlobalMembers.getEnumData(
      YGDirection::class, flags, directionOffset,
    )
  }

  fun directionBitfieldRef(): BitfieldRef<YGDirection> {
    return BitfieldRef(this, directionOffset)
  }

  fun flexDirection(): YGFlexDirection {
    return GlobalMembers.getEnumData(
      YGFlexDirection::class, flags, flexdirectionOffset,
    )
  }

  fun flexDirectionBitfieldRef(): BitfieldRef<YGFlexDirection> {
    return BitfieldRef(this, flexdirectionOffset)
  }

  fun justifyContent(): YGJustify {
    return GlobalMembers.getEnumData(
      YGJustify::class, flags, justifyContentOffset,
    )
  }

  fun justifyContentBitfieldRef(): BitfieldRef<YGJustify> {
    return BitfieldRef(this, justifyContentOffset)
  }

  fun alignContent(): YGAlign {
    return GlobalMembers.getEnumData(
      YGAlign::class, flags, alignContentOffset,
    )
  }

  fun alignContentBitfieldRef(): BitfieldRef<YGAlign> {
    return BitfieldRef(this, alignContentOffset)
  }

  fun alignItems(): YGAlign {
    return GlobalMembers.getEnumData(
      YGAlign::class, flags, alignItemsOffset,
    )
  }

  fun alignItemsBitfieldRef(): BitfieldRef<YGAlign> {
    return BitfieldRef(this, alignItemsOffset)
  }

  fun alignSelf(): YGAlign {
    return GlobalMembers.getEnumData(
      YGAlign::class, flags, alignSelfOffset,
    )
  }

  fun alignSelfBitfieldRef(): BitfieldRef<YGAlign> {
    return BitfieldRef(this, alignSelfOffset)
  }

  fun positionType(): YGPositionType {
    return GlobalMembers.getEnumData(
      YGPositionType::class, flags, positionTypeOffset,
    )
  }

  fun positionTypeBitfieldRef(): BitfieldRef<YGPositionType> {
    return BitfieldRef(this, positionTypeOffset)
  }

  fun flexWrap(): YGWrap {
    return GlobalMembers.getEnumData(
      YGWrap::class, flags, flexWrapOffset,
    )
  }

  fun flexWrapBitfieldRef(): BitfieldRef<YGWrap> {
    return BitfieldRef(this, flexWrapOffset)
  }

  fun overflow(): YGOverflow {
    return GlobalMembers.getEnumData(
      YGOverflow::class, flags, overflowOffset,
    )
  }

  fun overflowBitfieldRef(): BitfieldRef<YGOverflow> {
    return BitfieldRef(this, overflowOffset)
  }

  fun display(): YGDisplay {
    return GlobalMembers.getEnumData(
      YGDisplay::class, flags, displayOffset,
    )
  }

  fun displayBitfieldRef(): BitfieldRef<YGDisplay> {
    return BitfieldRef(this, displayOffset)
  }

  class BitfieldRef<T : Enum<T>>(
    val style: YGStyle,
    val offset: Int,
    private val enumValues: Array<T>,
  ) {
    fun getValue(enumClazz: KClass<T>): T {
      return GlobalMembers.getEnumData(enumClazz, enumValues, style.flags, offset)
    }

    fun setValue(x: T): BitfieldRef<T> {
      GlobalMembers.setEnumData(x::class, style.flags, offset, x)
      return this
    }
  }

  inline fun <reified T : Enum<T>> BitfieldRef(style: YGStyle, offset: Int) =
    BitfieldRef<T>(style, offset, enumValues())

  companion object {
    private const val directionOffset = 0
    private val flexdirectionOffset: Int =
      directionOffset + GlobalMembers.bitWidthFn<YGDirection>()
    private val justifyContentOffset: Int =
      flexdirectionOffset + GlobalMembers.bitWidthFn<YGFlexDirection>()
    private val alignContentOffset: Int =
      justifyContentOffset + GlobalMembers.bitWidthFn<YGJustify>()

    //  ~YGStyle() = default;
    private val alignItemsOffset: Int =
      alignContentOffset + GlobalMembers.bitWidthFn<YGAlign>()
    private val alignSelfOffset: Int =
      alignItemsOffset + GlobalMembers.bitWidthFn<YGAlign>()
    private val positionTypeOffset: Int =
      alignSelfOffset + GlobalMembers.bitWidthFn<YGAlign>()
    private val flexWrapOffset: Int =
      positionTypeOffset + GlobalMembers.bitWidthFn<YGPositionType>()
    private val overflowOffset: Int =
      flexWrapOffset + GlobalMembers.bitWidthFn<YGWrap>()
    private val displayOffset: Int =
      overflowOffset + GlobalMembers.bitWidthFn<YGOverflow>()
  }
}
